src/mds/SessionMap.h

```diff
@@ -255,28 +262,61 @@ class Session : public RefCountedObject {
 
 class MDS;
 
-class SessionMap {
+/**
+ * Encapsulate the serialized state associated with SessionMap.  Allows
+ * encode/decode outside of live MDS instance.
+ */
+class SessionMapStore {
+protected:
+    mds_rank_t rank;
 public:
-  MDS *mds;
-private:
   ceph::unordered_map<entity_name_t, Session*> session_map;
+  version_t version;
+
+  virtual void encode(bufferlist& bl) const;
+  virtual void decode(bufferlist::iterator& blp);
+  void dump(Formatter *f) const;
+
+  Session* get_or_add_session(const entity_inst_t& i) {
+    Session *s;
+    if (session_map.count(i.name)) {
+      s = session_map[i.name];
+    } else {
+      s = session_map[i.name] = new Session;
+      s->info.inst = i;
+      s->last_cap_renew = ceph_clock_now(g_ceph_context);
+    }
+
+    return s;
+  }
+
+  static void generate_test_instances(list<SessionMapStore*>& ls);
+
+  void reset_state()
+  {
+    session_map.clear();
+  }
+
+  SessionMapStore() : rank(MDS_RANK_NONE), version(0) {}
```

Comment:
I'm still concerned about this. Previously a SessionMap was a thing that belonged to an MDS and which nothing else touched. It was serialized for storage to disk but it was only modified by the MDS it belonged to, and a SessionMap for a rankless MDS was empty but that made sense — it had no sessions.

Now we're creating this SessionMapStore, whose purpose is to allow manipulation of dead MDSes. I get that we have a SessionMap for rankless MDSes, but I'm not very comfortable with giving it a data member which is modified only by the child SessionMap class via reference to its MDS, and which is logically part of its identity, but which doesn't influence anything.

Perhaps instead we can use the rank as a check about whether it's legal to add Sessions, serialize it with the rest of the data, and validate that the ranks match when doing a load()? (Or even just not create a SessionMap until we activate, but I'm not sure how hard that would be.)